secprofile/oci-profile.go

```diff
@@ -99,34 +95,56 @@ func (p *OCIProfile) AddNamespaces(nsTypes ...specs.LinuxNamespaceType) {
 	}
 }
 
-// AllowSyscallsWithArgs adds seccomp rules to allow syscalls with the given arguments if necessary
-func (p *OCIProfile) AllowSyscallsWithArgs(syscallsWithArgsToAllow map[types.Syscall][]specs.LinuxSeccompArg) {
-	defaultActError := p.OCI.Linux.Seccomp.DefaultAction == specs.ActErrno
+// RemoveNamespaces disables a list of namespaces
+func (p *OCIProfile) RemoveNamespaces(nsTypes ...specs.LinuxNamespaceType) {
+	for _, ns := range nsTypes {
+		for index, namespace := range p.OCI.Linux.Namespaces {
+			if namespace.Type == ns {
+				p.OCI.Linux.Namespaces = append(p.OCI.Linux.Namespaces[:index], p.OCI.Linux.Namespaces[index+1:]...)
+			}
+		}
+	}
+}
 
-	for syscallNameToAllow, syscallArgsToAllow := range syscallsWithArgsToAllow {
-		syscallNameToAllowStr := string(syscallNameToAllow)
+func allowSyscallWithArgs(seccompProfile *specs.LinuxSeccomp, syscallName types.Syscall, syscallArgs []specs.LinuxSeccompArg) *specs.LinuxSeccomp {
+	defaultActError := seccompProfile.DefaultAction == specs.ActErrno
 
-		for _, syscallRule := range p.OCI.Linux.Seccomp.Syscalls {
+	syscallNameToAllowStr := string(syscallName)
 
-			if syscallRule.Action == specs.ActAllow {
-				for _, syscallName := range syscallRule.Names {
-					if syscallName == syscallNameToAllowStr &&
-						((len(syscallArgsToAllow) == 0 && len(syscallRule.Args) == 0) ||
-							reflect.DeepEqual(syscallRule.Args, syscallArgsToAllow)) {
-						return
```

Comment:
This was a big bug, so here is the reason behind this change:
- when allowing a set of syscalls and args (a map so unordered set), we iterate over the set and try to create a seccomp rule allowing a syscall if necessary.
- if the whitelisting rule was already there for this syscall with those args we were leaving the function instead of doing the next iteration.. so we were randomly getting incomplete seccomp configuration because of iteration order on maps.

I broke down this function into two functions, one function `AllowSyscallsWithArgs ` iterating over the syscalls set to allow and a helper `allowSyscallWithArgs` function which whitelists 1 syscall with associated args if necessary.